File: Go/primes.go
// go build primes.go
// prime search trail division - no concurrency

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func IsPrime(n int) bool {
	if n < 0 {
		n = -n
	}
	switch {
	case n == 2:
		return true
	//case n < 2 || n%2 == 0: // we skip testing even numbers
	//return false
	default:
		for i := 3; i*i <= n; i += 2 {
			if n%i == 0 {
				return false
			}
		}
	}
	return true
}

func prompt(s string) string {
	fmt.Print(s + " ")
	reader := bufio.NewReader(os.Stdin)
	text, _ := reader.ReadString('\n')
	// convert CRLF to LF
	text = strings.Replace(text, "\n", "", -1)
	return text
}

func getInt(s string) int {
	temp, err := strconv.Atoi(s)
	if err == nil {
		return temp
	}
	return 0
}

func main() {
	var result []int
	var num = 0
	var text = ""

	fmt.Println("Looks for prime numbers from 1 to your input")

	for {
		result = nil
		num = 0
		text = prompt("Seek until what integer number?")
		num = getInt(text)

		start := time.Now()        // Start timer
		result = append(result, 2) // append 2 manually as we start searching from 3
		for i := 3; i < num; i += 2 {
			if IsPrime(i) {
				result = append(result, i)
			}
		}
		duration := time.Since(start) // End timer & duration
		elapsed := float64(duration) / float64(time.Millisecond)
		elapsed = elapsed / 1000.0
		fmt.Printf("Found %d primes.\n", len(result))
		fmt.Printf("Seconds: %.3f\n", elapsed)

		text = prompt("Print primes? (y/n)")
		if strings.Compare("y", text) == 0 {
			fmt.Printf("%v\n", result)
		}

		text = prompt("Another run? (y/n)")
		if strings.Compare("y", text) != 0 {
			break
		}
		fmt.Println("---------------------")
		fmt.Println("")
	}

}

File: Julia/primes.jl
#!/usr/bin/julia
#julia primes.jl

function Input(prompt)::String
    print(prompt)
    return chomp(readline())
end

function check_primes(limit)
	sq = limit^0.5
    for i in 3:2:trunc(Int, sq)
        limit%i != 0 || return false
    end
    return true
end

function prime_list(limit)
	result = Int64[]
	push!(result, 2)
	for x in 3:2:limit
		if check_primes(x)
			push!(result, x)
		end
	end
	result
end

function calc_primes(bound)
    @time p = prime_list(bound)
    print(length(p)," primes found\n")
end
 
while true
	println("*** Looks for primes ***")
	n = Input("Seek primes until what number? ")
	num = parse(Int64, n)
	
	calc_primes(num)
	
	resp = Input("Another run? (y/n) ")
	if resp != "y"
		break
	end
	println("")
end

